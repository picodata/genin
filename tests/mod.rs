use std::{
    fs::{read_to_string, File},
    io::Write,
    process::{Command, Output},
};

fn create_file(path: &str) {
    File::create(path).unwrap();
}

fn create_test_dir(path: &str) {
    std::fs::create_dir_all(path).expect("failed to create dir");
}

fn delete_test_dir(path: &str) {
    let _ = std::fs::remove_dir_all(path);
}

fn cleanup_test_dir(path: &str) {
    delete_test_dir(path);
    create_test_dir(path);
}

#[allow(unused)]
pub fn build_result_from_output(output: Output) -> String {
    let mut result = Vec::new();
    result.write_all(&output.stdout).unwrap();
    result.write_all(&output.stderr).unwrap();

    String::from_utf8(strip_ansi_escapes::strip(String::from_utf8(result).unwrap()).unwrap())
        .unwrap()
}

#[test]
fn genin_init() {
    cleanup_test_dir("tests/.genin_init");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("--output")
    .arg("tests/.genin_init/cluster.genin.yml")
    .output()
    .expect("Failed to execute command");

    let genin_init = build_result_from_output(output);
    let genin_init = format!(
        "{genin_init}\n{}",
        read_to_string("tests/.genin_init/cluster.genin.yml").unwrap()
    );

    insta::assert_display_snapshot!("genin_init", genin_init);
}

#[test]
fn warning_message_on_init_output() {
    cleanup_test_dir("tests/.warning_message_on_init_output");
    create_file("tests/.warning_message_on_init_output/cluster.genin.yml");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("-q")
    .current_dir("tests/.warning_message_on_init_output")
    .output()
    .expect("Failed to execute command");

    let mut result = Vec::new();
    result.write_all(&output.stdout).unwrap();
    result.write_all(&output.stderr).unwrap();

    println!("{}", String::from_utf8(result.clone()).unwrap());

    assert_eq!(
        result,
        b"WARN: the target file cluster.genin.yml already exists so the new file will \
            be saved with name cluster.genin.copy.yml\n"
    );
}

#[test]
fn genin_inspect() {
    cleanup_test_dir("tests/.genin_inspect");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("--output")
    .arg("tests/.genin_inspect/cluster.genin.yml")
    .output()
    .expect("Failed to execute command");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("inspect")
    .arg("--source")
    .arg("tests/.genin_inspect/cluster.genin.yml")
    .output()
    .expect("Failed to execute command");

    let genin_inspect = build_result_from_output(output);

    insta::assert_display_snapshot!("genin_inspect", genin_inspect);
}

#[test]
fn warning_message_on_build_output() {
    cleanup_test_dir("tests/.warning_message_on_build_output");
    create_file("tests/.warning_message_on_build_output/inventory.yml");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("-q")
    .current_dir("tests/.warning_message_on_build_output")
    .output()
    .expect("Failed to execute command");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("--source")
    .arg("cluster.genin.yml")
    .arg("-q")
    .current_dir("tests/.warning_message_on_build_output")
    .output()
    .expect("Failed to execute command");

    let mut result = Vec::new();
    result.write_all(&output.stdout).unwrap();
    result.write_all(&output.stderr).unwrap();

    println!("{}", String::from_utf8(result.clone()).unwrap());

    assert_eq!(
        result,
        b"WARN: the target file inventory.yml already exists so the new file will be \
            saved with name inventory.copy.yml\n"
    );
}

#[test]
fn init_with_comments() {
    cleanup_test_dir("tests/.init_with_comments");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("-q")
    .current_dir("tests/.init_with_comments")
    .output()
    .expect("Failed to execute command");

    let generated = std::fs::read_to_string("tests/.init_with_comments/cluster.genin.yml").unwrap();

    insta::assert_display_snapshot!(generated)
}

#[test]
fn build_from_state() {
    cleanup_test_dir("tests/.build_from_state");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("-s")
    .arg("tests/resources/cluster.genin.yml")
    .arg("-o")
    .arg("tests/.build_from_state/inventory.yml")
    .arg("--export-state")
    .arg("tests/.build_from_state/state.tgz")
    .output()
    .expect("Failed to execute command");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("-s")
    .arg("tests/.build_from_state/state.tgz")
    .arg("-o")
    .arg("tests/.build_from_state/inventory.yml")
    .arg("-f")
    .output()
    .expect("Failed to execute command");

    let inventory = read_to_string("tests/.build_from_state/inventory.yml").unwrap();

    insta::assert_display_snapshot!(inventory);

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("-s")
    .arg("tests/resources/cluster-new.genin.yml")
    .arg("-o")
    .arg("tests/.build_from_state/inventory_new.yml")
    .arg("--export-state")
    .arg("tests/.build_from_state/state.tgz")
    .output()
    .expect("Failed to execute command");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("-s")
    .arg("tests/.build_from_state/state.tgz")
    .arg("-o")
    .arg("tests/.build_from_state/inventory_new.yml")
    .arg("-f")
    .output()
    .expect("Failed to execute command");

    let inventory_new = read_to_string("tests/.build_from_state/inventory_new.yml").unwrap();

    insta::assert_display_snapshot!(inventory_new);
}

#[test]
fn sequential_upgrade_from_state() {
    cleanup_test_dir("tests/.sequential_upgrade_from_state");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("upgrade")
    .arg("--old")
    .arg("tests/resources/cluster.genin.yml")
    .arg("--new")
    .arg("tests/resources/cluster-new.genin.yml")
    .arg("--output")
    .arg("tests/.sequential_upgrade_from_state/v1_inventory.yml")
    .arg("--export-state")
    .arg("tests/.sequential_upgrade_from_state/v1_state.tgz")
    .arg("--state-dir")
    .arg("tests/.sequential_upgrade_from_state/.geninstate")
    .output()
    .expect("Failed to execute command");

    let cluster_to_cluster_new = build_result_from_output(output);

    let cluster_to_cluster_new = format!(
        "{cluster_to_cluster_new}\n{}",
        read_to_string("tests/.sequential_upgrade_from_state/v1_inventory.yml").unwrap()
    );

    insta::assert_display_snapshot!("cluster_to_cluster_new", cluster_to_cluster_new);

    ///////////////////////////////////////////////////////////////////////////////////////////////
    // upgrade from previous saved state
    ///////////////////////////////////////////////////////////////////////////////////////////////

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("upgrade")
    .arg("--old")
    .arg("tests/.sequential_upgrade_from_state/v1_state.tgz")
    .arg("--new")
    .arg("tests/resources/cluster-new-v2.genin.yml")
    .arg("--output")
    .arg("tests/.sequential_upgrade_from_state/v2_inventory.yml")
    .arg("--state-dir")
    .arg("tests/.sequential_upgrade_from_state/.geninstate")
    .output()
    .expect("Failed to execute command");

    let cluster_new_to_cluster_v2 = build_result_from_output(output);

    let cluster_new_to_cluster_v2 = format!(
        "{cluster_new_to_cluster_v2}\n{}",
        read_to_string("tests/.sequential_upgrade_from_state/v2_inventory.yml").unwrap()
    );

    insta::assert_display_snapshot!("cluster_new_to_cluster_v2", cluster_new_to_cluster_v2);

    ///////////////////////////////////////////////////////////////////////////////////////////////
    // upgrade from latest state
    ///////////////////////////////////////////////////////////////////////////////////////////////

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("upgrade")
    .arg("--from-latest-state")
    .arg("--new")
    .arg("tests/resources/cluster-new-v3.genin.yml")
    .arg("--output")
    .arg("tests/.sequential_upgrade_from_state/v3_inventory.yml")
    .arg("--state-dir")
    .arg("tests/.sequential_upgrade_from_state/.geninstate")
    .output()
    .expect("Failed to execute command");

    let cluster_v2_to_cluster_v3 = build_result_from_output(output);
    let cluster_v2_to_cluster_v3 = format!(
        "{cluster_v2_to_cluster_v3}\n{}",
        read_to_string("tests/.sequential_upgrade_from_state/v3_inventory.yml").unwrap()
    );

    insta::assert_display_snapshot!("cluster_v2_to_cluster_v3", cluster_v2_to_cluster_v3);
}

#[test]
fn sequential_upgrade_with_decreasing() {
    cleanup_test_dir("tests/.sequential_upgrade_with_decreasing");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("upgrade")
    .arg("--old")
    .arg("tests/resources/cluster-new-v3.genin.yml")
    .arg("--new")
    .arg("tests/resources/cluster-new-v4.genin.yml")
    .arg("--output")
    .arg("tests/.sequential_upgrade_with_decreasing/v1_inventory.yml")
    .arg("--state-dir")
    .arg("tests/.sequential_upgrade_with_decreasing/.geninstate")
    .output()
    .expect("Failed to execute command");

    let cluster_v3_to_cluster_v4 = build_result_from_output(output);

    let cluster_v3_to_cluster_v4 = format!(
        "{cluster_v3_to_cluster_v4}\n{}",
        read_to_string("tests/.sequential_upgrade_with_decreasing/v1_inventory.yml").unwrap()
    );

    insta::assert_display_snapshot!("cluster_v3_to_cluster_v4", cluster_v3_to_cluster_v4);

    ///////////////////////////////////////////////////////////////////////////////////////////////
    // upgrade from latest state
    ///////////////////////////////////////////////////////////////////////////////////////////////

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("upgrade")
    .arg("--from-latest-state")
    .arg("--new")
    .arg("tests/resources/cluster-new-v5.genin.yml")
    .arg("--output")
    .arg("tests/.sequential_upgrade_with_decreasing/v2_inventory.yml")
    .arg("--state-dir")
    .arg("tests/.sequential_upgrade_with_decreasing/.geninstate")
    .output()
    .expect("Failed to execute command");

    let cluster_v4_to_cluster_v5 = build_result_from_output(output);

    let cluster_v4_to_cluster_v5 = format!(
        "{cluster_v4_to_cluster_v5}\n{}",
        read_to_string("tests/.sequential_upgrade_with_decreasing/v2_inventory.yml").unwrap()
    );

    insta::assert_display_snapshot!("cluster_v4_to_cluster_v5", cluster_v4_to_cluster_v5);
}

#[test]
fn upgrade_consistency_100_times() {
    cleanup_test_dir("tests/.upgrade_consistency_100_times");

    for _ in 1..=100 {
        let output = Command::new(format!(
            "{}/target/debug/genin",
            std::env::var("CARGO_MANIFEST_DIR").unwrap()
        ))
        .arg("upgrade")
        .arg("--old")
        .arg("tests/resources/cluster.genin.yml")
        .arg("--new")
        .arg("tests/resources/cluster-new.genin.yml")
        .arg("--output")
        .arg("tests/.upgrade_consistency_100_times/inventory.yml")
        .arg("-f")
        .arg("--state-dir")
        .arg("tests/.upgrade_consistency_100_times/.geninstate")
        .output()
        .expect("Failed to execute command");

        let consistency_100_times = build_result_from_output(output);

        let consistency_100_times = format!(
            "{consistency_100_times}\n{}",
            read_to_string("tests/.upgrade_consistency_100_times/inventory.yml").unwrap()
        );

        insta::assert_display_snapshot!("consistency_100_times", consistency_100_times);
    }
}

#[test]
fn build_invalid_config() {
    cleanup_test_dir("tests/.build_invalid_config");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("-s")
    .arg("tests/resources/cluster-invalid.genin.yml")
    .arg("-o")
    .arg("tests/.build_invalid_config/inventory.yml")
    .output()
    .expect("Failed to execute command");

    let build_invalid_config = build_result_from_output(output);

    insta::assert_display_snapshot!(build_invalid_config);
}
