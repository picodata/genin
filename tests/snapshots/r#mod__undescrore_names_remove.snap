---
source: tests/mod.rs
assertion_line: 521
expression: result
---
+---------------------+---------------------+
|                  cluster                  |
+---------------------+---------------------+
|               datacenter-1                |
+---------------------+---------------------+
|      server-1       |      server-2       |
+---------------------+---------------------+
|   router_router-1   |   router_router-2   |
|   8081/3031         |   8081/3031         |
+---------------------+---------------------+
| storage_storage-1-1 | storage_storage-1-2 |
| 8082/3032           | 8082/3032           |
+---------------------+---------------------+
| storage_storage-2-1 | storage_storage-2-2 |
| 8083/3033           | 8083/3033           |
+---------------------+---------------------+
|     stateboard      |                     |
+---------------------+---------------------+

---
# List of replicasets as an array
topology:
  - name: router_router
    # How many masters we want, by default equal 1
    replicasets_count: 2
    # Array of roles for this instance
    roles:
      - router
      - failover-coordinator
  - name: storage_storage
    # How many masters we want, by default equal 1
    replicasets_count: 2
    # Number of replicas in replicaset, default 0
    replication_factor: 2
    # Array of roles for this instance
    roles:
      - storage
# List of regions, datacenters, and servers
hosts:
  - name: datacenter-1
    # Config with arbitrary key-values pairs
    config:
      # Specify http port to start counting from
      http_port: 8081
      # Specify binary port to start counting from
      binary_port: 3031
    # List of regions, datacenters, and servers
    hosts:
      - name: server-1
        # Config with arbitrary key-values pairs
        config:
          # Specify http port to start counting from
          http_port: 8081
          # Specify binary port to start counting from
          binary_port: 3031
          # Host or instance address (maybe IP or URI)
          address: 192.168.16.11
      - name: server-2
        # Config with arbitrary key-values pairs
        config:
          # Specify http port to start counting from
          http_port: 8081
          # Specify binary port to start counting from
          binary_port: 3031
          # Host or instance address (maybe IP or URI)
          address: 192.168.16.12
# Failover management options
failover:
  # Failover mode (stateful, eventual, disabled)
  mode: stateful
  # What is serve failover (stateboard, stateful)
  state_provider: stateboard
  # Timeout (in seconds), used by membership to mark `suspect` members as `dead`
  failover_timeout: 666
  # Abandon leadership when both the state provider quorum and at least one replica are lost (suitable in stateful mode only)
  fencing_enabled: true
  # Time (in seconds) to actuate fencing after the check fails
  fencing_timeout: 24
  # The period (in seconds) of performing the check
  fencing_pause: 42
  # Params for chosen in state_provider failover type
  stateboard_params:
    # Uri on which the stateboard will be available
    uri: "192.168.16.11:4401"
    # Stateboard password
    password: password
# Vars similar to those configured in the cartridge inventory
vars:
  # Username under which the ansible will connect to the servers
  ansible_user: ansible
  # Ansible user password
  ansible_password: ansible
  # Application name
  cartridge_app_name: myapp
  # Cookie for connecting to the administrative console of the instances
  cartridge_cluster_cookie: myapp-cookie
  # Path to the application package
  cartridge_package_path: /tmp/myapp.rpm
  # Indicates if vshard must be bootstrapped on the cluster
  cartridge_bootstrap_vshard: true

